### PR TITLE
Implementing overloaded unification

### PR DESCRIPTION
**Pull request by traiansf**
I've implemented unification; did not address matching.   strictness for lists still not working, but I'm not sure that is needed for WASM.

###### Reviewer checklist

- [ ] Test coverage: `stack test --coverage`
- [ ] Public API documentation: `stack haddock`
- [ ] Style conformance: `stylish-haskell`

---

### CODE REVIEW

**Comment by ttuegel** [comment]
I think that KWASM only uses overloaded `\rewrite` rules, so we can defer implementing this for matching.

I don't understand what you mean by,
> strictness for lists still not working

Could you expand on that?

File: src/main/k/working/tests/overloads/simple-exps/simple-exps.k [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+  syntax Val ::= Int
+  syntax Exp ::= Val
+  syntax Exp ::= Exp "+" Exp             [strict]
+  syntax Exp ::= "[" Exps "]"
+  syntax Exp ::= length(Exps)
+  syntax Exps ::= List{Exp,","}          [klabel(_,_)]  // automatically hybrid now
+  syntax Vals ::= List{Val,","}          [klabel(_,_)]
+  syntax Exps ::= Vals
+
+  rule [Vs:Vals] => length(Vs)
```
[the remark this answers is not in the document]
Comment (traiansf) [reply src/main/k/working/tests/overloads/simple-exps/simple-exps.k]: I wanted to make sure that the overloading simplification rules apply; the rule only applies when it's a list of values.

File: kore/src/Kore/Step/Simplification/AndTerms.hs [hunk 2]
```diff
@@ -935,6 +935,109 @@ constructorSortInjectionAndEquals
...
+    :: Ord variable
+    => SortedVariable variable
+    => Unparse variable
+    => GHC.HasCallStack
+    => Symbol
+    -> Symbol
+    -> Symbol
+    -> [TermLike variable]
+    -> TermLike variable
+revertOverloading
```
[the remark this answers is not in the document]
Comment (traiansf) [reply kore/src/Kore/Step/Simplification/AndTerms.hs]: correct, although it's virtually doing that. as detailed in the proposal, I'm just using the sorting information to recreate the lhs from the rhs.

**Comment by traiansf** [comment]
> I think that KWASM only uses overloaded `\rewrite` rules, so we can defer implementing this for matching.
> 
> I don't understand what you mean by,
> 
> > strictness for lists still not working
> 
> Could you expand on that?

I tried a more complex example, where lists were declared strict, which would create heating and cooling rules for them.  This seems to enter a cycle, but I'm not sure whether this is only due to overloading; it could also be due to how computing `isKResult` (or its negation) for complex datastructures such as lists.  Again, this might not be a problem for WASM as I believe lists are not strict there.

**Comment by ttuegel** [comment]
> This seems to enter a cycle, but I'm not sure whether this is only due to overloading; it could also be due to how computing `isKResult` (or its negation) for complex datastructures such as lists.

I agree, that doesn't sound like it's overloading-related. Could you open an issue for that?

File: kore/src/Kore/Step/Simplification/Overloading.hs [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    => GHC.HasCallStack
+    => TermSimplifier variable unifier
+    -> TermLike variable
+    -> TermLike variable
+    -> MaybeT unifier (Pattern variable)
+overloadedConstructorSortInjectionAndEquals
+    termMerger
+    first@(App_ firstHead _)
+    second@(App_ secondHead _)
+   = do
```
Comment (virgil-serbanuta) [inline kore/src/Kore/Step/Simplification/Overloading.hs]: should be indented 2, not 3 spaces.

**Pull request merged**